### PR TITLE
Fix NULL Pointer Read in Stateless Retry Scenario (on Server)

### PR DESCRIPTION
**Pull request by nibanks**
Fixes https://github.com/microsoft/msquic/issues/1938 . Essentially, if the server takes too long to set the configuration on the connection, it's possible for MsQuic to clean up some state that it still needs later (and ends up reading a NULL pointer).

The test code was added first, to 100% reproduce the original issue. Next the core code that cleans up the state was moved to the handshake complete instead of on first usage. Verified new tests now passed.

### CODE REVIEW

File: src/core/crypto.c [hunk 1]
```diff
@@ -1517,6 +1517,30 @@ QuicCryptoProcessTlsCompletion(
                 "Handshake confirmed (server)");
             QuicSendSetSendFlag(&Connection->Send, QUIC_CONN_SEND_FLAG_HANDSHAKE_DONE);
             QuicCryptoHandshakeConfirmed(&Connection->Crypto);
+
+            //
+            // Take this opportinuty to clean up the client chosen initial CID.
+            // It will be the second one in the list.
+            //
+            CXPLAT_DBG_ASSERT(Connection->SourceCids.Next != NULL);
```
Comment (anrossi) [inline src/core/crypto.c:1525]: should these be asserts? Or should they be runtime checks as well?

Comment (nibanks) [reply src/core/crypto.c:1525]: Asserts should be enough.

**Review by anrossi — APPROVED** [review]

**Comment by qzhuyan** [comment]
I can confirm cherry-pick d62ee62 to v1.6.0 solved issue in #1938

**Comment by nibanks** [comment]
> I can confirm cherry-pick [d62ee62](https://github.com/microsoft/msquic/commit/d62ee623ded780bb311d21c17f244715d15edd30) to v1.6.0 solved issue in #1938

Great! Thanks @qzhuyan!

**Pull request merged**